File: frontend/src/app/features/cuentas/cuentas.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';

/** saldo/monto llegan como string exacto — ver backend/app/schemas/comunes.py::Dinero */
export interface CuentaOut {
  cuenta_id: number;
  numero_cuenta: string;
  cci: string;
  tipo_cuenta: 'ahorro' | 'corriente';
  moneda: 'PEN' | 'USD';
  saldo: string;
  fecha_apertura: string;
  estado: string;
}

export interface SaldoOut {
  cuenta_id: number;
  numero_cuenta: string;
  moneda: string;
  saldo: string;
}

export interface MovimientoOut {
  transaccion_id: number;
  fecha_hora: string;
  tipo: string;
  monto: string;
  canal: string;
  estado: string;
  concepto: string | null;
  direccion: 'entrada' | 'salida';
  cuenta_origen_id: number | null;
  cuenta_destino_id: number | null;
  cuenta_origen_numero: string | null;
  cuenta_destino_numero: string | null;
}

export interface MovimientosPagina {
  items: MovimientoOut[];
  total: number;
  pagina: number;
  tamano: number;
}

export interface ComisionRetiroOut {
  monto: string;
  retiros_gratis_restantes: number | null;
}

@Injectable({ providedIn: 'root' })
export class Cuentas {
  private readonly http = inject(HttpClient);

  listar() {
    return this.http.get<CuentaOut[]>(`${environment.apiUrl}/cuentas`);
  }

  saldo(cuentaId: number) {
    return this.http.get<SaldoOut>(`${environment.apiUrl}/cuentas/${cuentaId}/saldo`);
  }

  comisionRetiro(cuentaId: number) {
    return this.http.get<ComisionRetiroOut>(`${environment.apiUrl}/cuentas/${cuentaId}/comision-retiro`);
  }

  movimientos(cuentaId: number, opts: { desde?: string; hasta?: string }, pagina = 1, tamano = 20) {
    let params = new HttpParams().set('pagina', pagina).set('tamano', tamano);
    if (opts.desde) params = params.set('desde', opts.desde);
    if (opts.hasta) params = params.set('hasta', opts.hasta);
    return this.http.get<MovimientosPagina>(`${environment.apiUrl}/cuentas/${cuentaId}/movimientos`, { params });
  }
}

File: backend/app/routers/cuentas.py
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.core.security import require_cliente
from app.schemas.cuentas import CuentaOut, SaldoOut
from app.schemas.transacciones import MovimientoOut, MovimientosPagina
from app.services import cuentas as cuentas_service

router = APIRouter(prefix="/cuentas", tags=["cuentas"])


@router.get("", response_model=list[CuentaOut])
def listar(db: Session = Depends(get_db), cliente_id: int = Depends(require_cliente)):
    return cuentas_service.listar(db, cliente_id)


@router.get("/{cuenta_id}/saldo", response_model=SaldoOut)
def saldo(cuenta_id: int, db: Session = Depends(get_db), cliente_id: int = Depends(require_cliente)):
    cuenta = cuentas_service.obtener_propia(db, cliente_id, cuenta_id)
    return SaldoOut(cuenta_id=cuenta.cuenta_id, numero_cuenta=cuenta.numero_cuenta, moneda=cuenta.moneda, saldo=cuenta.saldo)


@router.get("/{cuenta_id}/movimientos", response_model=MovimientosPagina)
def movimientos(
    cuenta_id: int,
    desde: date | None = None,
    hasta: date | None = None,
    pagina: int = Query(1, ge=1),
    tamano: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    cliente_id: int = Depends(require_cliente),
):
    filas, total, numeros = cuentas_service.movimientos(db, cliente_id, cuenta_id, desde, hasta, pagina, tamano)
    items = [
        MovimientoOut(
            transaccion_id=t.transaccion_id, fecha_hora=t.fecha_hora, tipo=t.tipo, monto=t.monto,
            canal=t.canal, estado=t.estado, direccion=direccion,
            cuenta_origen_id=t.cuenta_origen_id, cuenta_destino_id=t.cuenta_destino_id,
            cuenta_origen_numero=numeros.get(t.cuenta_origen_id), cuenta_destino_numero=numeros.get(t.cuenta_destino_id),
        )
        for t, direccion in filas
    ]
    return MovimientosPagina(items=items, total=total, pagina=pagina, tamano=tamano)
